File: src/lib/work-order.ts
import type { Expense } from "@/lib/types";

/** AppFolio-style work order: e.g. 76-2234 */
const WORK_ORDER_PATTERN = /^(\d{2})-(\d{3,5})$/;

const WORK_ORDER_EXTRACT_PATTERNS = [
  /\b(?:wo|w\/o|work\s*order)\s*#?\s*(\d{2}-\d{3,5})\b/i,
  /\b(\d{2}-\d{3,5})\b/,
];

function isLikelyDateFalsePositive(prefix: string, suffix: string): boolean {
  const month = Number(prefix);
  const year = Number(suffix);
  return month >= 1 && month <= 12 && year >= 2020 && year <= 2035;
}

/** Normalize user or AI input to AppFolio-style xx-xxxx, or null if invalid. */
export function normalizeWorkOrderNumber(value: unknown): string | null {
  if (value == null || value === "") return null;

  let cleaned = String(value).trim().toUpperCase();
  cleaned = cleaned.replace(/^(?:WO|W\/O)\s*#?\s*/i, "");

  const strict = cleaned.match(WORK_ORDER_PATTERN);
  if (strict) {
    const [, prefix, suffix] = strict;
    if (isLikelyDateFalsePositive(prefix, suffix)) return null;
    return `${prefix}-${suffix}`;
  }

  for (const pattern of WORK_ORDER_EXTRACT_PATTERNS) {
    const match = cleaned.match(pattern);
    if (match?.[1]) {
      const normalized = normalizeWorkOrderNumber(match[1]);
      if (normalized) return normalized;
    }
  }

  return null;
}

export function extractWorkOrderFromText(text: string): string | null {
  if (!text.trim()) return null;

  for (const line of text.split(/\r?\n/)) {
    for (const pattern of WORK_ORDER_EXTRACT_PATTERNS) {
      const match = line.match(pattern);
      if (match?.[1]) {
        const normalized = normalizeWorkOrderNumber(match[1]);
        if (normalized) return normalized;
      }
    }
  }

  return null;
}

export function resolveWorkOrderNumber(
  explicit: unknown,
  ...textSources: Array<string | null | undefined>
): string | null {
  const fromExplicit = normalizeWorkOrderNumber(explicit);
  if (fromExplicit) return fromExplicit;

  for (const source of textSources) {
    if (!source) continue;
    const found = extractWorkOrderFromText(source);
    if (found) return found;
  }

  return null;
}

export function formatWorkOrderLabel(
  workOrderNumber: string | null | undefined,
): string {
  if (!workOrderNumber) return "No work order";
  return `WO ${workOrderNumber}`;
}

export function expenseNeedsWorkOrder(
  expense: Pick<Expense, "billableStatus">,
): boolean {
  return expense.billableStatus === "billable";
}

export function isWorkOrderMissing(
  expense: Pick<Expense, "billableStatus" | "workOrderNumber">,
): boolean {
  return expenseNeedsWorkOrder(expense) && !expense.workOrderNumber;
}

export function countMissingWorkOrders(expenses: Expense[]): number {
  return expenses.filter(isWorkOrderMissing).length;
}

File: agent/src/kai_agent/services/work_order.py
"""Work order parsing — port of src/lib/work-order.ts."""

from __future__ import annotations

import re

WORK_ORDER_PATTERN = re.compile(r"^(\d{2})-(\d{3,5})$")

WORK_ORDER_EXTRACT_PATTERNS = [
    re.compile(r"\b(?:wo|w/o|work\s*order)\s*#?\s*(\d{2}-\d{3,5})\b", re.I),
    re.compile(r"\b(\d{2}-\d{3,5})\b"),
]


def _is_likely_date_false_positive(prefix: str, suffix: str) -> bool:
    month = int(prefix)
    year = int(suffix)
    return 1 <= month <= 12 and 2020 <= year <= 2035


def normalize_work_order_number(value: object | None) -> str | None:
    if value is None or value == "":
        return None

    cleaned = str(value).strip().upper()
    cleaned = re.sub(r"^(?:WO|W/O)\s*#?\s*", "", cleaned, flags=re.I)

    strict = WORK_ORDER_PATTERN.match(cleaned)
    if strict:
        prefix, suffix = strict.group(1), strict.group(2)
        if _is_likely_date_false_positive(prefix, suffix):
            return None
        return f"{prefix}-{suffix}"

    for pattern in WORK_ORDER_EXTRACT_PATTERNS:
        match = pattern.search(cleaned)
        if match and match.group(1):
            normalized = normalize_work_order_number(match.group(1))
            if normalized:
                return normalized

    return None


def extract_work_order_from_text(text: str) -> str | None:
    if not text.strip():
        return None

    for line in text.splitlines():
        for pattern in WORK_ORDER_EXTRACT_PATTERNS:
            match = pattern.search(line)
            if match and match.group(1):
                normalized = normalize_work_order_number(match.group(1))
                if normalized:
                    return normalized

    return None


def resolve_work_order_number(explicit: object | None, *text_sources: str | None) -> str | None:
    from_explicit = normalize_work_order_number(explicit)
    if from_explicit:
        return from_explicit

    for source in text_sources:
        if not source:
            continue
        found = extract_work_order_from_text(source)
        if found:
            return found

    return None
